curp/src/server/cmd_board.rs

```diff
@@ -133,18 +125,38 @@ impl<C: Command> CommandBoard<C> {
         self.shutdown_notifier.notify(usize::MAX);
     }
 
+    /// Get an execution result
+    fn get_er(&self, id: &ProposeId) -> Option<Result<C::ER, C::Error>> {
+        self.er_buffer.get(id).cloned()
+    }
+
+    /// Get an execution result and an after sync result
+    #[allow(clippy::type_complexity)] // it is easy to understand
+    fn get_er_asr(
+        &self,
+        id: &ProposeId,
+    ) -> (
+        Option<Result<C::ER, C::Error>>,
+        Option<Result<C::ASR, C::Error>>,
+    ) {
+        (
+            self.er_buffer.get(id).cloned(),
+            self.asr_buffer.get(id).cloned(),
+        )
+    }
+
     /// Wait for an execution result
+    #[allow(clippy::expect_used)]
     pub(super) async fn wait_for_er(
         cb: &CmdBoardRef<C>,
         id: &ProposeId,
-    ) -> Result<C::ER, C::Error> {
-        loop {
-            if let Some(er) = cb.map_read(|cb_r| cb_r.er_buffer.get(id).cloned()) {
-                return er;
-            }
-            let listener = cb.write().er_listener(id);
-            listener.await;
+    ) -> Option<Result<C::ER, C::Error>> {
+        if let Some(er) = cb.map_read(|cb_r| cb_r.get_er(id)) {
```

Comment:
could we just remove L154-L156?

Reply:
The listener will only be notified during the insert result. If a wait sync request misses a notification once, it will continue to wait at L158